pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -1681,3 +1691,56 @@ func isDeviceOpened(deviceToDetach AttachedVolume, hostUtil hostutil.HostUtils)
 	}
 	return deviceOpened, nil
 }
+
+func cancelControlPlaneExpansion(pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume, client clientset.Interface) (bool, error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.RecoverVolumeExpansionFailure) {
+		return false, nil
+	}
+
+	newSize := pvc.Spec.Resources.Requests[v1.ResourceStorage]
+	statusSize := pvc.Status.Capacity[v1.ResourceStorage]
+	pvSize := pv.Spec.Capacity[v1.ResourceStorage]
+	if statusSize.Cmp(newSize) >= 0 && pvSize.Cmp(newSize) >= 0 {
+		klog.Infof("PVC %s already is of requested size", util.GetPersistentVolumeClaimQualifiedName(pvc))
+		err := util.MarkResizeFinished(pvc, newSize, client)
+		if err != nil {
+			detailedErr := fmt.Errorf("Error marking pvc %s as resized : %v", util.GetPersistentVolumeClaimQualifiedName(pvc), err)
+			return true, detailedErr
+		}
+		return true, nil
+	}
+	return false, nil
+}
+
+func cancelNodeExpansion(pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume, client clientset.Interface) (bool, error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.RecoverVolumeExpansionFailure) {
+		return false, nil
+	}
+
+	// Just to be safe we should only cancel node expansion of volumes which already have
+	// filesystem resize pending condition
+	if !hasFileSystemResizePendingCondition(pvc) {
+		return false, nil
+	}
+
+	pvSpecCap := pv.Spec.Capacity[v1.ResourceStorage]
+	// if we are here that means PVC status already is of size greater or equal to
+	// PV capacity and hence no node expansion is necessary.
+	err := util.MarkFSResizeFinished(pvc, pvSpecCap, client)
```

Comment:
Is it possible for PVC status to be greater than PV capacity?  I thought PVC status is updated last.  If PVC status is bigger than PV capacity, which one is the real size?

Reply:
In general `pvc.Status.Capacity` should not be greater than `pv.Spec.Capacity` but from volume expansion code flow perspective if that happens then there is no volume expansion needed on the node and expansion operation can be marked as finished.